From a2c69372644f9bc36e0ddb36b74118673c09ade2 Mon Sep 17 00:00:00 2001
From: Donald Sharp <sharpd@nvidia.com>
Date: Tue, 10 Mar 2026 12:22:14 -0400
Subject: [PATCH] zebra: Do not clear selected flag on route about to be
 imported

When creating a new re to import into a different table, do
not blindly copy the selected or use RR distance flag(s) over
to the new re.  This is because that route has undergone
it's own rib processing already and resettting it shows wrong
information in the `show ip route table X` command since
the routes will loose the fact that they have been selected.

Signed-off-by: Donald Sharp <sharpd@nvidia.com>
---
 zebra/redistribute.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/zebra/redistribute.c b/zebra/redistribute.c
index d3aa77ba4..b336ce974 100644
--- a/zebra/redistribute.c
+++ b/zebra/redistribute.c
@@ -723,6 +723,7 @@ int zebra_add_import_table_entry(struct zebra_vrf *zvrf, safi_t safi, struct rou
 	struct nexthop_group *ng;
 	route_map_result_t ret = RMAP_PERMITMATCH;
 	afi_t afi;
+	uint32_t import_flags;
 
 	afi = family2afi(rn->p.family);
 	if (rmap_name)
@@ -731,7 +732,6 @@ int zebra_add_import_table_entry(struct zebra_vrf *zvrf, safi_t safi, struct rou
 							 rmap_name);
 
 	if (ret != RMAP_PERMITMATCH) {
-		UNSET_FLAG(re->flags, ZEBRA_FLAG_SELECTED);
 		zebra_del_import_table_entry(zvrf, safi, rn, re);
 		return 0;
 	}
@@ -749,15 +749,15 @@ int zebra_add_import_table_entry(struct zebra_vrf *zvrf, safi_t safi, struct rou
 			break;
 	}
 
-	if (same) {
-		UNSET_FLAG(same->flags, ZEBRA_FLAG_SELECTED);
+	if (same)
 		zebra_del_import_table_entry(zvrf, safi, rn, same);
-	}
 
-	UNSET_FLAG(re->flags, ZEBRA_FLAG_RR_USE_DISTANCE);
+	import_flags = re->flags;
+	UNSET_FLAG(import_flags, ZEBRA_FLAG_SELECTED);
+	UNSET_FLAG(import_flags, ZEBRA_FLAG_RR_USE_DISTANCE);
 
-	newre = zebra_rib_route_entry_new(0, ZEBRA_ROUTE_TABLE, re->table, re->flags, re->nhe_id,
-					  zvrf->table_id, re->metric, re->mtu,
+	newre = zebra_rib_route_entry_new(0, ZEBRA_ROUTE_TABLE, re->table, import_flags,
+					  re->nhe_id, zvrf->table_id, re->metric, re->mtu,
 					  zebra_import_table_distance[afi][safi][re->table],
 					  re->tag);
 
-- 
2.54.0

